### DISCUSSION TITLE
mini.pairs: auto-insert semicolons?

### DISCUSSION THREAD

**Discussion by mrjones2014** [nvim-mini/mini.nvim#956] [Q&A] [answered]
I'd like to auto-insert semicolons in Nix code, e.g. when the code is like

```nix
{
  some_thing = 
}
```

when I type a `{` at the end of the line, I'd like it to expand to the following, where `|` is my cursor position.

```nix
{
  some_thing = { | };
}
```

I tried this in a `ftplugin` for Nix, but this puts my cursor between the `}` and the `;` instead of between `{}`:

```nix
local mp = require('mini.pairs')

mp.map_buf(0, 'i', '{', { action = 'open', pair = '{};', register = { cr = false } })
mp.map_buf(0, 'i', '}', { action = 'close', pair = '{};', register = { cr = false } })
```

If possible, I'd also like to _not_ insert the semicolon if the `{}` is inside a list type, like

```nix
{
  things = [
    # since this is inside square brackets, don't add the ;
    { }
  ];
}
```

**Comment by echasnovski** [comment, marked as answer]
I am afraid, this is not possible with 'mini.pairs':
- Both "open" and "close" parts  of a pair are expected to be a single character. More than that [is not supported](https://github.com/echasnovski/mini.nvim/blob/f1dd64cf2ac1b348bd440e59446abf680c7188d6/doc/mini-pairs.txt#L19-L20). So no "};" as "close" part is supported.
- Condition on whether a pair should be included can be made [only on two character neighborhood](https://github.com/echasnovski/mini.nvim/blob/f1dd64cf2ac1b348bd440e59446abf680c7188d6/doc/mini-pairs.txt#L9-L10). So no conditions like "it is inside a list type" are supported.
